File: central/laboratory/avalon-net/Composition/ModelDefault/DefaultContainmentModelExportHelper.cs
/* 
* Copyright 2003-2004 The Apache Software Foundation
* Licensed  under the  Apache License,  Version 2.0  (the "License");
* you may not use  this file  except in  compliance with the License.
* You may obtain a copy of the License at 
* 
*   http://www.apache.org/licenses/LICENSE-2.0
* 
* Unless required by applicable law or agreed to in writing, software
* distributed  under the  License is distributed on an "AS IS" BASIS,
* WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
* implied.
* 
* See the License for the specific language governing permissions and
* limitations under the License.
*/
namespace Apache.Avalon.Composition.Model.Default
{
	using System;

	using Apache.Avalon.Meta;
	using Apache.Avalon.Composition.Data;
	
	/// <summary> A utility class that assists in the location of a model relative
	/// a supplied path.
	/// 
	/// </summary>
	/// <author>  <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
	/// </author>
	/// <version>  $Revision: 1.1 $ $Date: 2004/02/28 22:15:42 $
	/// </version>
	class DefaultContainmentModelExportHelper
	{
		//-------------------------------------------------------------------
		// static
		//-------------------------------------------------------------------
		
		//-------------------------------------------------------------------
		// immutable state
		//-------------------------------------------------------------------
		
		private IContainmentContext m_context;
		private IContainmentModel m_model;
		
		//-------------------------------------------------------------------
		// constructor
		//-------------------------------------------------------------------
		
		public DefaultContainmentModelExportHelper(IContainmentContext context, IContainmentModel model)
		{
			m_context = context;
			m_model = model;
		}
		
		//-------------------------------------------------------------------
		// implementation
		//-------------------------------------------------------------------
		
		public virtual DefaultServiceModel[] createServiceExport()
		{
			
			ServiceDirective[] export = m_context.ContainmentProfile.ExportDirectives;
			DefaultServiceModel[] services = new DefaultServiceModel[export.Length];
			for (int i = 0; i < export.Length; i++)
			{
				ServiceDirective service = export[i];
				System.Type type = getServiceExportClass(service);
				IDeploymentModel provider = locateImplementionProvider(service);
				services[i] = new DefaultServiceModel(service, type, provider);
			}
			return services;
		}
		
		/// <summary> Return the class declared by a container service export declaration.</summary>
		/// <returns> the exported service interface class
		/// </returns>
		/// <exception cref=""> ModelException if the class cannot be resolved
		/// </exception>
		private System.Type getServiceExportClass(ServiceDirective service)
		{
			return service.Reference.Type;
		}
		
		/// <summary> Given a service directive declared by a container, locate a model 
		/// with this containment model to map as the provider.  If not model
		/// is explicity declared, the implementation will attempt to construct
		/// a new model based on packaged profiles and add the created model to
		/// the set of models within this container.
		/// 
		/// </summary>
		/// <param name="service">the service directive
		/// </param>
		/// <returns> the implementing deployment model
		/// </returns>
		/// <exception cref=""> ModelException if an implementation is not resolvable 
		/// </exception>
		private IDeploymentModel locateImplementionProvider(ServiceDirective service)
		{
			String path = service.Path;
			if (null != path)
			{
				IDeploymentModel provider = m_model.GetModel(path);
				
				if (null == provider)
				{
					//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
					//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
					String error = "Implemention provider path [" + path + "] for the exported service [" + service.Reference + "] in the containment model " + m_model + " does not reference a known model.";
					throw new ModelException(error);
				}
				else
				{
					return provider;
				}
			}
			else
			{
				//UPGRADE_NOTE: Final was removed from the declaration of 'dependency '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				DependencyDescriptor dependency = new DependencyDescriptor("export", service.Reference, null);
				
				//UPGRADE_NOTE: Final was removed from the declaration of 'repository '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				IModelRepository repository = m_context.ModelRepository;
				//UPGRADE_NOTE: Final was removed from the declaration of 'candidates '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				IDeploymentModel[] candidates = repository.Models;
				//UPGRADE_NOTE: Final was removed from the declaration of 'selector '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				IModelSelector selector = new DefaultModelSelector();
				IDeploymentModel provider = selector.Select(candidates, dependency);
				if (null != provider)
				{
					return provider;
				}
				else
				{
					ITypeRepository repo = m_context.TypeLoaderModel.TypeRepository;
					DeploymentProfile[] profiles = repo.GetProfiles(dependency, false);

					if (profiles.Length != 0)
					{
						return m_model.AddModel(profiles[0]);
					}
					else
					{
						//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
						//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
						String error = "Could not locate a provider for the exported service [" + dependency.Service + "] in the containment model " + m_model;
						throw new ModelException(error);
					}
				}
			}
		}
	}
}

File: avalon/composition/impl/src/java/org/apache/avalon/composition/model/impl/DefaultContainmentModelExportHelper.java
/* 
 * Copyright 2004 Apache Software Foundation
 * Licensed  under the  Apache License,  Version 2.0  (the "License");
 * you may not use  this file  except in  compliance with the License.
 * You may obtain a copy of the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed  under the  License is distributed on an "AS IS" BASIS,
 * WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
 * implied.
 * 
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.avalon.composition.model.impl;

import org.apache.avalon.composition.data.ServiceDirective;
import org.apache.avalon.composition.data.DeploymentProfile;
import org.apache.avalon.composition.model.ContainmentModel;
import org.apache.avalon.composition.model.ContainmentContext;
import org.apache.avalon.composition.model.DeploymentModel;
import org.apache.avalon.composition.model.ModelException;
import org.apache.avalon.composition.model.ModelRepository;
import org.apache.avalon.composition.model.ModelSelector;
import org.apache.avalon.composition.model.TypeRepository;

import org.apache.avalon.excalibur.i18n.ResourceManager;
import org.apache.avalon.excalibur.i18n.Resources;

import org.apache.avalon.meta.info.DependencyDescriptor;

/**
 * A utility class that assists in the location of a model relative
 * a supplied path.
 *
 * @author <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
 * @version $Revision: 1.1 $ $Date: 2004/02/07 20:21:03 $
 */
class DefaultContainmentModelExportHelper
{
    //-------------------------------------------------------------------
    // static
    //-------------------------------------------------------------------

    private static final Resources REZ =
      ResourceManager.getPackageResources( 
        DefaultContainmentModelExportHelper.class );

    //-------------------------------------------------------------------
    // immutable state
    //-------------------------------------------------------------------

    private final ContainmentContext m_context;
    private final ContainmentModel m_model;

    //-------------------------------------------------------------------
    // constructor
    //-------------------------------------------------------------------

    public DefaultContainmentModelExportHelper( 
      ContainmentContext context, ContainmentModel model )
    {
        m_context = context;
        m_model = model;
    }

    //-------------------------------------------------------------------
    // implementation
    //-------------------------------------------------------------------

    public DefaultServiceModel[] createServiceExport() throws ModelException
    {

        ServiceDirective[] export = 
          m_context.getContainmentProfile().getExportDirectives();
        DefaultServiceModel[] services = new DefaultServiceModel[ export.length ];
        for( int i=0; i<export.length; i++ )
        {
            ServiceDirective service = export[i];
            Class clazz = getServiceExportClass( service );
            DeploymentModel provider = 
              locateImplementionProvider( service );
            services[i] = 
              new DefaultServiceModel( service, clazz, provider ); 
        }
        return services;
    }

   /**
    * Return the class declared by a container service export declaration.
    * @return the exported service interface class
    * @exception ModelException if the class cannot be resolved
    */
    private Class getServiceExportClass( ServiceDirective service )
      throws ModelException
    {
        String classname = service.getReference().getClassname();
        try
        {
            ClassLoader classloader = m_context.getClassLoader();
            return classloader.loadClass( classname );
        }
        catch( Throwable e )
        {
            final String error = 
              "Cannot load service class [" 
              + classname 
              + "].";
            throw new ModelException( error, e );
        }
    }

   /**
    * Given a service directive declared by a container, locate a model 
    * with this containment model to map as the provider.  If not model
    * is explicity declared, the implementation will attempt to construct
    * a new model based on packaged profiles and add the created model to
    * the set of models within this container.
    * 
    * @param service the service directive
    * @return the implementing deployment model
    * @exception ModelException if an implementation is not resolvable 
    */
    private DeploymentModel locateImplementionProvider( ServiceDirective service )
      throws ModelException
    {
        final String path = service.getPath();
        if( null != path )
        {
            DeploymentModel provider = m_model.getModel( path );
            if( null == provider )
            {
                final String error = 
                  "Implemention provider path [" 
                  + path 
                  + "] for the exported service [" 
                  + service.getReference()
                  + "] in the containment model "
                  + m_model
                  + " does not reference a known model.";
               throw new ModelException( error );
            }
            else
            {
                return provider;
            }
        }
        else
        {
            final DependencyDescriptor dependency = 
              new DependencyDescriptor( 
                "export", 
                service.getReference() );

            final ModelRepository repository = m_context.getModelRepository();
            final DeploymentModel[] candidates = repository.getModels();
            final ModelSelector selector = new DefaultModelSelector();
            DeploymentModel provider = selector.select( candidates, dependency );
            if( null != provider )
            {
                return provider;
            }
            else
            {
                TypeRepository repo = 
                  m_context.getClassLoaderModel().getTypeRepository();
                DeploymentProfile profile = 
                  repo.getProfile( dependency, false );
                if( profile != null )
                {
                    return m_model.addModel( profile );
                }
                else
                {
                    final String error = 
                      "Could not locate a provider for the exported service [" 
                        + dependency.getReference()
                        + "] in the containment model "
                        + m_model;
                    throw new ModelException( error );
                }
            }
        }
    }

}
